CPLAT-9775 DOCS: New boilerplate docs / migration guide (#475)
Workiva/over_react: doc/new_boilerplate_migration.md

* __[New Boilerplate Updates](#new-boilerplate-updates)__
    * [Constraints](#constraints)
    * [Updates](#updates)
        * [Remove Annotations](#remove-annotations)
        * [Ignore Ungenerated Warnings Project-Wide](#ignore-ungenerated-warnings-project-wide)
        * [Use Mixin-Based Props Declaration that Disallows Subclassing](#use-mixin-based-props-declaration-that-disallows-subclassing)
* __[Function Components](#function-components)__
    * [Constraints](#function-component-constraints)
    * [Design](#design)
* __[Upgrading](#upgrading)__

## Background

While converting the boilerplate for OverReact component declaration from Dart 1 
transformers to Dart 2 builders, we encountered several constraints that made us 
choose between dropping backwards compatibility (mainly support for props class 
inheritance), and a less-than-optimal boilerplate.

To make the Dart 2 transition as smooth as possible, we chose to keep the new 
boilerplate version as backwards-compatible as possible, while compromising the 
cleanliness of the boilerplate. In time, we found that this wasn't great from a 
user experience or from a tooling perspective.

> Knowing this, we now have the opportunity to implement an improved version of 

Reviewer: ```suggestion
Knowing this, and having dropped support for Dart 1, we now have the opportunity to implement an improved version of 
```

OverReact boilerplate that fixes issues introduced in the latest version, as 
well as other miscellaneous ones.

### Function Component Boilerplate

React function components, along with [hooks](https://reactjs.org/docs/hooks-intro.html), 
are used heavily in modern React component architecture, and we would like to 
take advantage of their benefits within the Workiva ecosystem. However, to be 
truly valuable, we need strongly typed prop APIs for these components, which 
means we must provide a way to declare function components using OverReact 

> boilerplate.

Reviewer: Since this isn't part of the "migration", I'm not sure this belongs here.  If we want to keep it here - it should go in some sort of section talking about "new boilerplate features" or something like that.

Reviewer: ```suggestion
### Function Component Boilerplate

React function components, along with [hooks](https://reactjs.org/docs/hooks-intro.html), 
are used heavily in modern React component architecture, and we want consumers to 
be able to take advantage of their benefits. However, to be 
truly valuable, we need strongly typed prop APIs for these components, which 
means we must provide a way to declare function components using OverReact 
boilerplate.
```

### Problems with Previous Boilerplate

#### Public API and Code Generation

The current, Dart-2-only boilerplate generates public props classes:

```dart
// User-authored
@Props()
class _$FooProps extends BarProps {
  String foo;
}

// Generated in .over_react.g.dart
class FooProps extends _$FooProps with _$FooPropsAccessorsMixin {   
  static const PropsMeta meta = ...;
  ... 
}
```

In using the build packages's `build_to: cache` to generate public APIs, command-line 
tools like `dartanalyzer` and `dartdoc` don't function properly for packages that do this. 

This pattern can also degrade the dev experience, by requiring a build before code is 
statically valid, and also requiring rebuilds in some cases to consume public API 
updates during their development (e.g., writing new component props classes).

> The transitional boilerplate, currently used in almost all repos, does not have these 

Reviewer: ```suggestion
The transitional _(backwards-compatible with Dart 1)_ boilerplate, currently used in almost all repos, does not have these 
```

    main() method of all consuming apps' entrypoints.
    
    2. a user-authored entrypoint (field initializer, method invocation, 
    constructor, etc.) that imports (or pulls in via a part) and references 
    generated code (what we have now).

#### Self-Imposed Constraints:

1. Keep component declarations as terse and user-friendly as possible.

2. Use `build_to: cache` (for more information, see: 
[pkg:build docs](https://github.com/dart-lang/build/blob/master/docs/builder_author_faq.md#when-should-a-builder-build-to-cache-vs-source)).
    
    `build_to:cache` should be used when generated code is dependent on the 
    library's underlying implementation. This may not be strictly the case 
    today, but if we commit to `build_to: cache`, we will have more 
    flexibility in the future to make improvements or fix bugs to OverReact 
    code generation without requiring a (very expensive) breaking change. 
    
    It would also result in improvements to the builder being propagated 
    immediately as soon as they're consumed by wdesk, as opposed to having 
    to regenerate code and release within every consumer library.

> 3. Make source code statically analyzable without running codegen.

Reviewer: ```suggestion
3. Make source code statically analyzable without running a build.
```

    The build docs instruct not to use build_to: cache to generate public 
    APIs, and command-line tools like `dartanalyzer` and `dartdoc` don't 
    function properly for packages that do this. 
    
    Generating public APIs can also degrade the dev experience, by requiring 
    a build before code is statically valid, and also requiring rebuilds in 
    some cases to consume public API updates during their development (e.g., 
    writing new component props classes).
    
4. Provide some means of sharing props/state declarations between 
components.

    Being able to share props/state between multiple components is useful, 
    especially when composing them together. We also have many legacy 
    components that currently share props, and want to make it possible 
    to upgrade them.
    
5. Provide a simple migration path for _most_ components in the Wdesk 

> ecosystem.

Reviewer: ```suggestion
5. Provide a simple migration path for _most_ components in our consumer 
ecosystems.
```

    We can support new/old boilerplate at the same time, and slowly phase 

>     out the old as we migrate over to it using codemods.

Reviewer: ```suggestion
    out the old as we migrate over to it using [our `boilerplate_upgrade` codemod](#upgrading).
```

    
    For cases that don't migrate cleanly, we can use the Wdesk versioning 
    policy to replace them with APIs that use the new boilerplate in major 
    versions or using versioned APIs. This applies to a small set of 
    components within Wdesk whose props classes are public APIs, 

>     concentrated in web_skin_dart and to alesser extent graph_ui.

Reviewer: ```suggestion
    For cases that don't migrate cleanly within the Workiva ecosystem, we can use the Wdesk versioning 
    policy to replace them with APIs that use the new boilerplate in major 
    versions or using versioned APIs.
```

    
6. Only support Component2 components.

    The builder has different code paths for Component/Component2, and 
    supporting an additional boilerplate for both would increase code 
    complexity and effort needed to build/test it. 

### Updates

#### Remove Annotations

`@Factory()`, `@Props()` and `@Component()` annotations add additional 
visual clutter to the boilerplate, and are redundant since the factory/
props/component declarations already have a consistent/restricted 
structure and naming scheme that makes it clear to the builder parsing 
logic that a component is being defined, and what each part is.

```diff
- @Factory()
UiFactory<FooProps> Foo =
    // ignore: undefined_identifier
    $Foo;

- @Props()
class _$FooProps extends BarProps {
  String foo;
}

- @Component2()
class FooComponent extends UiComponent2<FooProps> {
  @override
  render() => 'foo: ${props.foo}';
}
```

> Annotations could still be used, opt-in, if custom configuration is needed.

Reviewer: ```suggestion
Annotations can still be used, opt-in, if custom configuration is needed.
```

```dart
@Props(keyNamespace: 'customNamespace.')
class _$FooProps extends BarProps {
  String foo;
}
```

#### Ignore Ungenerated Warnings Project-Wide

Right now, we have to add `// ignore: uri_has_not_been_generated` to each 
component library on the part/import that references generated code.

Ignoring this hint globally within analysis_options.yaml:

```yaml
 analyzer:
   errors:
     uri_has_not_been_generated: ignore 
```

Allows individual ignores to be omitted, which will reduce clutter in 
the component boilerplate.

```diff
- // ignore: uri_has_not_been_generated
part 'foo.over_react.g.dart';
```

#### Use Mixin-Based Props Declaration that Disallows Subclassing

_Constraints_:

* Props classes must directly subclass UiProps, only inheriting other props 
via mixins.

    * This requires consumers to include every single mixin within their `with` 
    clause, allowing the builder to mix in the generated code corresponding 
    to those mixins.
    
* Generated props mixin classes (`$FooPropsMixin`) must be exported along with 
their source class `$FooProps`. These APIs are considered public, but should 

> only ever be referenced from other code generated by the over_react builder.

Reviewer: Not sure I understand this.  Is this something the consumer needs to do / understand, or is it an implementation detail?

##### New Boilerplate (from Dart 2 only boilerplate)

> 

Reviewer: I would say - since the vast majority of consumers are currently using the transitional boilerplate, that we show that.

}
```

_Generated code_:

```dart
part of 'foo.dart';

//
// (Component and factory code is pretty much the same.)
//

mixin $FooPropsMixin on FooPropsMixin {
  @override
  String get foo => props['foo'];
  @override
  set foo(String value) => props['foo'] = value;
}

class _$FooPropsImpl extends FooProps
    // These mixins are derived from the list of mixins in the source
    with $FooPropsMixin, $BarPropsMixin {}
```

##### Abbreviated Version

When no other mixins are used, skip the props class and just use the mixin.

```dart
import 'package:over_react/over_react.dart';

part 'foo.over_react.g.dart';

UiFactory<FooProps> Foo = $Foo; // ignore: undefined_identifier

mixin FooProps on UiProps {
  String foo;
}

class FooComponent extends UiComponent2<FooProps> {
  @override
  render() => 'foo: ${props.foo}';
}
```

##### Props Meta Changes

Props meta will be generated as an overridden getter on the component 
as opposed to the current static field, and will allow similar access 
of prop keys as before.

This eliminates the current `meta` portion of the boilerplate which has 
to reference more generated code.

Prop meta from all mixins can be accessed, allowing us to default 

> consumedProps to all props statically accessible from that component.

Reviewer: ```suggestion
`consumedProps` to all props statically accessible from that component.
```

```diff
import 'package:over_react/over_react.dart';

+ part 'foo.over_react.g.dart';

- class FooPropsMapView extends UiPropsMapView with SomeOtherPropsMixin {
-   FooPropsMapView(Map backingMap) : super(backingMap);
- }

+ UiFactory<FooMapViewProps> FooMapView = $FooMapView; // ignore: undefined_identifier

+ class FooMapViewProps = UiProps with SomeOtherPropsMixin;

usage() {
-   final mapView = FooPropsMapView(someExistingMap);
+   final mapView = FooMapView(someExistingMap);
}
```

## Function Components

### Function Component Constraints 

Includes all of [the constraints listed in the Boilerplate Updates section](#constraints), 
ignoring parts about backwards-compatibility.

* Should be as visually uncluttered as possible.

* Should not wrap excessively for longer component names.

* Should be easy to transition between having and not having default 
props, and boilerplate shouldn't change shape drastically when doing so.

* Function calls using generated functions should be avoided since they 
don't allow generic type inference of the `props` arg in the function 
closure.

### Design

```dart
import 'package:over_react/over_react.dart';

part 'foo.over_react.g.dart';

> UiFactory<FooProps> Foo = uiFunctionComponent(

Reviewer: I thought we decided against LHS typing?

```suggestion
final Foo = uiFunctionComponent<FooProps>(
```

Author: I got this from the current RFD. Greg said we need the LHS typing because otherwise it won't be statically typed in some repos.

Reviewer: @greglittlefield-wf is this the case?

Reviewer 2: Yeah, the idea is that LHS typing
1. keeps things consistent with non-function components, which need the LHS typing to statically analyze since the RHS is `dynamic` before it's generated
2. is compatible with the type_annotate_public_apis Effective Dart lint

across repos that have

  (props) {
    return 'foo: ${props.foo}'; 
  },
  $fooPropsConfig, // ignore: undefined_identifier
); 

mixin FooProps on UiProps {
  String foo;
}
```

Here, `uiFunctionComponent` gets a generic parameter of `FooProps` inferred 

> from the LHS typing, allowing props to be statically typed as `FooProps`.

Reviewer: See above... I thought we were opting for no LHS typing?

The generated `$FooPropsConfig` is passed in as an argument, and serves 
as the entrypoint to the generated code. 

#### With Default Props

`defaultProps` on function components is 
[already deprecated](https://github.com/facebook/react/pull/16210).

> Use null-aware operators to default null values. This provides almost the

Reviewer: ```suggestion
Instead, we use null-aware operators to default null values. This provides almost the
```

same behavior as `defaultProps`, but with the restriction that a given prop
__must either be nullable or have a default value, but not both__.

```dart

> UiFactory<FooProps> Foo = uiFunctionComponent(

Reviewer: ```suggestion
final Foo = uiFunctionComponent<FooProps>(
```

  (props) {
    final foo = props.foo ?? 'default foo value';

    return 'foo: $foo'; 
  },
  $fooPropsConfig, // ignore: undefined_identifier
); 
```

#### With propTypes

```dart

> UiFactory<FooProps> Foo = uiFunctionComponent(

Reviewer: ```suggestion
final Foo = uiFunctionComponent<FooProps>(
```

// be declared in this file. 
UiFactory<FooProps> createFooHoc(UiFactory otherFactory) { 
  Object closureVariable; 
  // ... 

  final FooHoc = uiFunctionComponent<FooProps>(
    (props) { 
      return otherFactory()( 
        Dom.div()('closureVariable: ${closureVariable}'), 
        Dom.div()('prop foo: ${props.foo}'), 
      );   
    },
    $fooPropsConfig, // ignore: undefined_identifier 
  ); 

  return FooHoc; 
} 
```

## Upgrading

To update your repository to the new boilerplate, you can use 
[over_react_codemod](https://github.com/Workiva/over_react_codemod)'s 
`boilerplate_upgrade` executable to make it easier. This codemod goes 
through the repository and updates the boilerplate as necessary. While 
the codemod will handle many basic updates, it will still need to be 
supplemented with some manual checks and refactoring. 

Before running the codemod, run `semver_audit` inside your repository 
and save the report using the following commands:

1. `pub global activate semver_audit --hosted-url=https://pub.workiva.org`
2. `pub global run semver_audit generate 2> semver_report.json`

This will allow the codemod to check whether or not components are

> public API.

Reviewer: ```suggestion
If you are migrating a Workiva library, before running the codemod, run `semver_audit` inside your repository 
and save the report using the following commands:

1. `pub global activate semver_audit --hosted-url=https://pub.workiva.org`
2. `pub global run semver_audit generate 2> semver_report.json`

This will allow the codemod to check whether or not components are
public API.

If you are migrating a library outside of the Workiva ecosystem, run the command 
below with the `--treat-all-components-as-private` flag.
```

Then, run the codemod by following the directions within 

> [the executable](https://github.com/Workiva/over_react_codemod/blob/master/lib/src/executables/boilerplate_upgrade.dart#L32)

Author: This link won't work until the boilerplate_upgrade PR is merged into master.

> while inside your repository.

Reviewer: ```suggestion
from the root of your local copy of the repository.
```
